src/invariant/OrderBookIsNotCrossed.cpp

```diff
@@ -0,0 +1,230 @@
+{
+    auto const priceA = priceAsDouble(a.price);
+    auto const priceB = priceAsDouble(b.price);
+
+    if (priceA < priceB)
+    {
+        return true;
+    }
+    if (priceA == priceB)
+    {
+        // We order non-passive offers before passive offers so that we
+        // can find any non-passive offer at a given price at the front
+        // of the list of offers at that price.
+        if ((a.flags & PASSIVE_FLAG) != (b.flags & PASSIVE_FLAG))
+        {
+            return (a.flags & PASSIVE_FLAG) == 0;
+        }
+        return a.offerID < b.offerID;
+    }
+    return false;
+}
+
+void
+OrderBookIsNotCrossed::updateOrderBook(LedgerTxnDelta const& ltxd)
+{
+    for (auto const& entry : ltxd.entry)
+    {
+        // there are three possible "deltas" for an offer:
+        //      CREATED:  (nil)     -> LedgerKey
+        //      MODIFIED: LedgerKey -> LedgerKey
+        //      DELETED:  LedgerKey -> (nil)
+        if (entry.first.type() == InternalLedgerEntryType::LEDGER_ENTRY &&
+            entry.first.ledgerKey().type() == OFFER)
+        {
+            if (mRestoreBeforeNextUpdate)
+            {
+                mOrderBook = mOrderBookSnapshot;
+                mRestoreBeforeNextUpdate = false;
+            }
+            if (entry.second.previous)
+            {
+                auto const& oe =
+                    entry.second.previous->ledgerEntry().data.offer();
+                mOrderBook[{oe.selling, oe.buying}].erase(oe);
+            }
+            if (entry.second.current)
+            {
+                auto const& oe =
+                    entry.second.current->ledgerEntry().data.offer();
+                mOrderBook[{oe.selling, oe.buying}].emplace(oe);
+            }
+        }
+    }
+}
+
+std::string
+OrderBookIsNotCrossed::check(AssetPairSet const& assetPairs)
+{
+    assert(!mRestoreBeforeNextUpdate);
+    for (auto const& assetPair : assetPairs)
+    {
+        auto checkCrossedResult =
+            checkCrossed(assetPair.first, assetPair.second, mOrderBook);
+        if (!checkCrossedResult.empty())
+        {
+            return checkCrossedResult;
+        }
+    }
+
+    return std::string{};
+}
+
+std::shared_ptr<Invariant>
+OrderBookIsNotCrossed::registerAndEnableInvariant(Application& app)
+{
+    auto invariant =
+        app.getInvariantManager().registerInvariant<OrderBookIsNotCrossed>();
+    app.getInvariantManager().enableInvariant(invariant->getName());
+    return invariant;
+}
+
+std::string
+OrderBookIsNotCrossed::getName() const
+{
+    return "OrderBookIsNotCrossed";
+}
+
+std::string
+OrderBookIsNotCrossed::checkOnOperationApply(Operation const& operation,
+                                             OperationResult const& result,
+                                             LedgerTxnDelta const& ltxDelta)
+{
+    updateOrderBook(ltxDelta);
+    auto assetPairs = extractAssetPairs(ltxDelta);
+    return assetPairs.size() > 0 ? check(assetPairs) : std::string{};
```

Comment:
I *think* that the reason things are super slow is that `assetPairs.size() > 0` is always true even when there are no changes to the order book from this operation. In fact we should not perform any work in `extractAssetPairs` if the operation didn't touch the order book.

Reply:
Hmm, why must `assetPairs.size()` be `> 0`?  It looks to me as though if there are no `OFFER`s in `LedgerTxnDelta`, then it will be `0`, and I put in some logs and they say that it is often `0`.

One thing I mentioned to Jon about performance which I should bring up is that exec speed of fuzzing seems highly variable on my laptop, even from run to run without code changes (unlike stability, for example, which is generally within a fraction of a percent across many runs).  Fuzzing does make my laptop fan roar, and I have a suspicion (though no confirmation) that there might be highly variable amounts of CPU-throttling going on depending on I-don't-know-what thermal factors.

Reply:
meh and I thought my head was straight in the morning :)

yeah this is extracting only from the current txn object so not it